### ISSUE TITLE
Dependabot timing out after Yarn Berry upgrade

### ISSUE THREAD

**Issue by bmulholland** [dependabot/dependabot-core#5978]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

yarn berry

### Package manager version

3.2.4

### Language version

Node 16

### Manifest location and content before the Dependabot update

/package.json

### dependabot.yml content

# To get started with Dependabot version updates, you'll need to specify which
# package ecosystems to update and where the package manifests are located.
# Please see the documentation for all configuration options:
# https://help.github.com/github/administering-a-repository/configuration-options-for-dependency-updates

version: 2
updates:
  - package-ecosystem: "npm"
    directory: "/"
    schedule:
      interval: "weekly"
      day: "monday"

  - package-ecosystem: "github-actions"
    directory: "/"
    schedule:
      interval: "weekly"
      day: "monday"

### What you expected to see, versus what you actually saw

I upgraded to yarn berry after v2 support was GA'd, and dependabot hasn't generated a PR since. Instead, it times out after over an hour running. The log is so long that it struggles to load in my browser.

https://github.com/recitalsoftware/recital-frontend/network/updates/492103567

Dependabot encountered an unknown error
Dependabot failed to update your dependencies because an unexpected error occurred. See the logs for more details.

A couple excerpts from the log that seem relevant:

updater | <job_492103567> ➤ YN0058: │ @nuxtjs/apollo@https://github.com/recitalsoftware/apollo-module.git#commit=424994eb55baf5e94526a4d52f21b83f45fcc8e7: Packing the package failed (exit code 1, logs can be found here: /tmp/xfs-cc70de15/pack.log)
updater | <job_492103567> ➤ YN0000: └ Completed in 20s 713ms
updater | <job_492103567> ➤ YN0000: Failed with errors in 24s 439ms
updater | ERROR <job_492103567> /home/dependabot/common/lib/dependabot/shared_helpers.rb:314:in `run_shell_command'

And later:

updater | time="2022-10-26T10:45:41Z" level=warning msg="timeout running job" error="waiting for updater: waiting for container: context deadline exceeded" job_id=492103567

**Comment by pavera** [comment]
@bmulholland thanks for taking the time to submit this issue.  I took a brief look and it does appear to be something around those lines you mentioned.  In order to get access to the logs in `/tmp/<random dir>/pack.log` and further debug the issue I'd need access to your repo.  If you're willing/able to grant that access we'd need a support ticket created via https://support.github.com. The support engineer will be able to request access on our behalf and I'd be happy to dig deeper into this issue with you.

**Comment by bmulholland** [comment]
Sure, filed #1849996

**Comment by 6f6d6172** [comment]
We're seeing similar, but for Bundler. Repositories that used to complete in a reasonable amount of time are now just timing out.

Notice the timestamps: fetching package metadata is no issue but eventually it just... stops, until the container's killed for timing out ~30m later.
```
...
  proxy | 2022/10/27 16:19:48 [516] GET https://index.rubygems.org:443/info/powerbar
  proxy | 2022/10/27 16:19:48 [516] 200 https://index.rubygems.org:443/info/powerbar
  proxy | 2022/10/27 16:19:48 [518] GET https://index.rubygems.org:443/info/ansi
  proxy | 2022/10/27 16:19:48 [518] 200 https://index.rubygems.org:443/info/ansi
updater | time="2022-10-27T16:49:30Z" level=info msg="task complete" container_id=job-493175820-updater exit_code=137 job_id=493175820 step=updater
updater | time="2022-10-27T16:49:31Z" level=warning msg="timeout running job" error="waiting for updater: waiting for container: context deadline exceeded" job_id=493175820
```

It appears we can't manually trigger dependabot runs any more either (`Insights -> Dependency Graph -> Dependabot`); the UI claims we haven't configured dependabot yet, but when I click the button to do so, it takes me to a screen that shows a fully configured `dependabot.yml` file.

Let me know if you think this is a distinct problem; I can open a separate issue. I've also filed a ticket, #1850965.

**Comment by pavera** [comment]
@6f6d6172 your issue is probably a distinct issue. If you have a support ticket already our support escalation process should handle it shortly.

**Comment by bmulholland** [comment]
FYI, I've granted support access now

**Comment by pavera** [comment]
Thanks @bmulholland. I've been able to run this in the debugger and it looks like the issue is possibly in corepack when it tries to pack a "Yarn Classic" package from source.  We use a proxy to isolate credentials from untrusted code execution, and corepack appears to be hard coded to reach out to https://classic.yarnpkg.com to install yarn v1 and perform the pack operation.  When this happens the request fails as it is not using our proxy.

I will investigate further and open an issue with corepack once I've fully isolated the issue.

Here are the additional logs I was able to gather:
```
Packing @nuxtjs/apollo@https://github.com/recitalsoftware/apollo-module.git#commit=424994eb55baf5e94526a4d52f21b83f45fcc8e7 from sources
<job_cli> ➤ YN0000: │ /tmp/xfs-e6ba7108 STDOUT Using Yarn Classic for bootstrap. Reason: "__metadata" key not found in yarn.lock, must be a Yarn classic lockfile
<job_cli> ➤ YN0000: │ /tmp/xfs-e6ba7108 STDOUT 
<job_cli> ➤ YN0000: │ /tmp/xfs-e6ba7108 STDOUT ➤ YN0000: Retrieving https://classic.yarnpkg.com/latest.js
<job_cli> ➤ YN0000: │ /tmp/xfs-e6ba7108 STDOUT ➤ YN0001: RequestError: getaddrinfo EAI_AGAIN classic.yarnpkg.com
```

**Comment by bmulholland** [comment]
Thanks, I've tried that but the job is still timing out. I did get one PR upgrade just now, so the basics are working, but the job seems to take too long to run. Either the timeout needs to be increased, or whatever is slowing down the job needs to be fixed.

**Comment by pavera** [comment]
I'm curious how long it takes for you to run `yarn install` on the repo? That is all we're doing during the job, and it does seem to take a long time on the first dependency because of the large number of dependencies.  Once that first install is complete and we've cached the dependencies we get through subsequent dependencies in around 4 minutes each.  It seems the errors you're seeing now are possibly due to conflicting requirements holding back updates.
```
Error processing office-addin-debugging (Dependabot::NpmAndYarn::FileUpdater::NoChangeError)
updater | ERROR <job_496980145> No files were updated!
```
As far as the timeout in general, for large projects with a large number of dependencies timeouts are expected to happen. In this case we generate as many PRs as possible with the time we're given. We attempt to make things as fast as possible, and I'll definitely keep reviewing this to see if we can speed things up, but we are to some extent at the mercy of how performant the underlying ecosystem tooling is.

**Comment by bmulholland** [comment]
`yarn install` takes 4s on the repo. If I clobber `node_modules` _and_ clean the yarn cache, it takes exactly 4 minutes on my system. Before the yarn berry upgrade, dependabot was running just fine. I checked its output fairly often for other reasons and never once saw a timeout error. I don't think the project is especially large? It's a pretty standard Nuxt app with a few extra plugins, eslint & plugins, cypress, jest, storybook, and that's pretty much it.

I'm not sure what you say about conflicting requirements, either; our app has very little requirements on package versions. Most packages are set to allow upgrades to `latest`. To pick the example you chose, a local `yarn up core-js` upgrades to the latest version of that package in 13s, changing only `core-js` and flagging no conflicting requirements. I don't understand why dependabot would say there's no updates in that case; it seems like a clear bug?

**Comment by pavera** [comment]
I've identified a couple of places where we can speed things up significantly, however, they are at the expense of yarn zero-installs working correctly so we have to be careful to implement the update in a clean way for those that have zero-installs enabled. I'm working on a larger PR to fix this.

Some of the updates are not being created on your repo because somehow `yarn install` is detecting a CI environment in your repo and enabling immutable installs. I've got a fix for this case nearly ready to test.

Additionally, it seems we've got a bug when we attempt to run `yarn up <packagename>@latest`. This does not update the lockfile in your repo for some reason. Testing locally with a minimal project/lockfile I'm unable to reproduce this behavior, but in the debugger on your repo if I specify `@latest` on any `yarn up` command it does not update the lockfile to the new version.  If I leave the version off completely eg. `yarn up core-js` then the lockfile is updated correctly. This appears to be expected yarn behavior somehow, though I don't know how to structure package.json/yarn.lock to recreate it.

**Comment by bmulholland** [comment]
Thanks, @pavera. It looks like there may be several things going on.

1. I had an env var set in Github to force CI=true. IIRC, I did this because the PRs generated by dependabot were running my tests without that env var set, which caused them to fail. I've tried removing this to see what happens.
2. Even without the CI option, the build still times out. Note that I do plan to switch to zero-installs in future (currently blocked because some packages I use are still working on their support). If dependabot can't handle zero-installs and a project this size with current settings, maybe the time budget needs to be increased?
3. Independently of that, package upgrades seem not to be working, even locally. I can `y up core-js`, and that works. However, just `y up`, or `y up core-js@latest`, has no effect, as you say. It looks like Yarn may have changed some behaviours about this in v2/v3, though I have yet to find any reference to it.

Note that # 2 is what I opened this issue for, and it may be a separate issue from (and red herring for) # 3

**Comment by pavera** [comment]
Having the repo configured with zero-installs is currently faster than non-zero install configured repos due to having a fully populated cache in the repo. This greatly speeds up `yarn install`. 

I'm working on https://github.com/dependabot/dependabot-core/pull/6017 to address the performance issues we can by running in update-lockfile mode when the repo is not configured for zero-installs, this prevents yarn from downloading/installing all of the dependencies in the project for each update attempt. The remaining slow step I've identified is in peer dependency resolution, this step currently downloads metadata and tgz files for each dependency and takes 3-4 minutes per update. We don't have a proposed fix here yet but I'm looking at it next (after I get the above PR merged).

The CI option and # 3 just cause us to waste time running an update that ultimately doesn't result in a PR, with both of those issues fixed we'll produce more PRs in the currently allotted time though we'll probably still timeout until at least the above PR is deployed, and probably until we address the peer dependency resolution.

**Comment by bmulholland** [comment]
That all sounds spot on. Thanks, @pavera.

**Comment by pavera** [comment]
I've now deployed https://github.com/dependabot/dependabot-core/pull/6017.  I'll now be working on a small update to fix point # 3.

**Comment by bmulholland** [comment]
Yeah it's so strange. That command does indeed do what I want, and it's perfect for Dependabot (as the docs note!). But there's no local equivalent: if I want to upgrade something on my system, I want to install it, too. This seems like a pretty major change in approach in Yarn, and they didn't even write about it? I'm just confused.

**Comment by bmulholland** [comment]
This time it created two PRs, but still timed out. Checking the logs, it failed to update packages that did have an update. I can reproduce this locally, e.g. `yarn up @cypress/webpack-preprocessor@latest --mode=update-lockfile` has no effect. However, a pure `yarn up @cypress/webpack-preprocessor` upgrades the package successfully.

This is really damaging my opinion of yarn. What a mess of inconsistent, unexplained behaviour. Thanks for dealing with it.

**Comment by pavera** [comment]
Thanks for working through this with me.  I also wonder why we're seeing different behavior as locally I'm unable to reproduce the exact behavior in test repos, but in your repo I definitely saw the same behavior that any attempt with `@latest` produced no effect on the lockfile. I'm going to move on to investigating the peer dependency resolution speed for now as I don't think we want to be updating package.json which is apparently the only way we can get a lockfile update.

**Comment by bmulholland** [comment]
@pavera Quick update here: I've been talking about it on the Yarn discord (you can join from their home page). The key takeaway so far is that, with Yarn Modern/Berry/v2+, _if the package.json version value doesn't change, yarn won't update it_. This applies to `latest`, but also any version range that doesn't include the full package version (e.g. "~1.3.5" works, while "<5" won't be upgraded).

FWIW, In my view this is a serious change in how Yarn works, and should be clearly documented at minimum. I'm also concerned about the security implications of people who think they are getting updates from Yarn but actually silently aren't. Maybe I'm missing something though, or over-reacting... what do you think?

Specifically, to quote https://github.com/arcanis:

> basically what happens is that Yarn sees that your dependency range is already latest, which is already resolved, so it doesn't re-resolve it (you can see yarn up as setting the dependency range to the given string for all workspaces at the same time) 
> when you don't put @latest on the command-line, Yarn needs to find out what's the version on the registry to know what it should put as range (which ends up being a semver version, ie different from the literal "latest" string)
> we probably should reset the resolution when using yarn up (we already do in yarn add), but it's unclear to me why you need to put the @latest in your CLI command
> if the range before and after is the same (and latest is a range, just like ^1.0.0, beta or whatever-dist-tag-i-can-think-of), the current behavior of yarn up is to assume that there's nothing to do. It doesn't have anything to do with latest, it's just how the command currently handles one edge case.

**Comment by pavera** [comment]
I've put up #6077 which speeds things up a bit when doing peer dependency checks, I'm still concerned it's too slow but it is a marked improvement. As far as your research and comment thank you! They mention that yarn add does reset the resolution, maybe I'll investigate using that. 

>  but it's unclear to me why you need to put the @latest in your CLI command

Is the confusion here why we would use `yarn up xyz@latest` at all and recommending that we run `yarn up xyz` instead? We can do that, but it then changes the version constraint in `package.json`. If that is the correct path I can finish up my PR that implements this change. At least people will start getting updates which to your point about security is the main concern.

**Comment by bmulholland** [comment]
I don't really understand what the intent is, TBH. For instance, as it stands, setting ranges like "<5.0.0" will never produce an update with `yarn up` or `yarn up package@<5.0.0`. The only way to upgrade is with `yarn up package` -- which will ignore the "<5" entirely. So actually, with that in mind, I don't think you can do that approach without upgrading too aggressively.

I'd suggest jumping on the Discord and trying to ask what's best. I tried to make my case there but don't think I succeeded in communicating the importance or urgency.

**Comment by pavera** [comment]
I caught up with the discussion on discord, seems the `-R` flag may be the ticket, I'll start testing that now.

**Comment by pavera** [comment]
Had a chance to do some testing today. It looks like if the version in package.json starts with a non-digit ie 'latest', '^', '~', '<' I should use `-R`, if it starts with a digit I need to run without `-R`. I'll have a PR up shortly to implement this logic. Let me know if you can think of places where this logic falls down? Basically if there is a resolution that needs to take place, I need to run with `-R`, if there is an exact version then I have to run without `-R` to get the latest version, running with `-R` when no resolution is necessary results in a noop.

**Comment by bmulholland** [comment]
Oh, actually, maybe the use of `-R` should also take into account the [versioning strategy](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#versioning-strategy)? Use `-R` when versioning-strategy is `lockfile-only`, and no `-R` (so it updates package.json) when the versioning strategy isn't that?

**Comment by pavera** [comment]
I've finally deployed https://github.com/dependabot/dependabot-core/pull/6077 which should speed up the slowest part of the job. I believe the versioning strategy is handled earlier in the process so I think by the time it gets to updating `yarn.lock` we can actually just use `-R` I'll be adding some tests and hopefully getting this deployed today as well.

**Comment by bmulholland** [comment]
Sounds good. I've downgraded to Yarn 1 for now, and will upgrade again when the changes are out so I can confirm the fix for you.

**Comment by pavera** [comment]
I've now merged the `-R` update so I'm closing this issue as I believe all the points identified are now addressed. Please re-open if it turns out we've got more to do.

**Comment by bmulholland** [comment]
I've upgraded back to Yarn Berry (without PnP) and confirmed that the updates are both working and not timing out. Thanks for all your work and help here, @pavera!

**Comment by jeffwidman** [comment]
Thanks for circling back... yeah, if you're still seeing issues but only tangentially related to this issue please file a new issue.
